analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs

```diff
@@ -28,21 +28,45 @@ internal static class Binary
         public static ProgramState Process(SymbolicContext context, IBinaryOperationWrapper binary) =>
             context.State[binary.LeftOperand] is { } left
             && context.State[binary.RightOperand] is { } right
-            && left.HasConstraint<BoolConstraint>()
-            && right.HasConstraint<BoolConstraint>()
-            && BinaryBoolConstraint(binary.OperatorKind, left.HasConstraint(BoolConstraint.True), right.HasConstraint(BoolConstraint.True)) is { } newConstraint
+            && BinaryConstraint(binary.OperatorKind, left, right) is { } newConstraint
                 ? context.SetOperationConstraint(newConstraint)
                 : context.State;
 
-        private static SymbolicConstraint BinaryBoolConstraint(BinaryOperatorKind kind, bool left, bool right) =>
-            kind switch
+        private static SymbolicConstraint BinaryConstraint(BinaryOperatorKind kind, SymbolicValue left, SymbolicValue right)
+        {
+            if (left.HasConstraint<BoolConstraint>() && right.HasConstraint<BoolConstraint>())
+            {
+                return BinaryBoolConstraint(kind, left.HasConstraint(BoolConstraint.True), right.HasConstraint(BoolConstraint.True));
+            }
+            else if (left.HasConstraint<ObjectConstraint>() && right.HasConstraint<ObjectConstraint>())
+            {
+                return BinaryNullConstraint(kind, left.HasConstraint(ObjectConstraint.Null), right.HasConstraint(ObjectConstraint.Null));
+            }
+            else
```

Comment:
We should also support this matrix:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/boolean-logical-operators#nullable-boolean-logical-operators

Are we following the lifting rules?
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-value-types#lifted-operators

These could be follow-ups.

Reply:
The lifting will be most likely part of #4250 that is part of our sprint

Reply:
Okay. I left a comment there to not lose the links to the C# spec.